src/docmanager/index.ts

```diff
@@ -0,0 +1,408 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import {
+  IContentsModel, IKernelId, IKernelSpecId, IContentsOpts, IKernel,
+  INotebookSession
+} from 'jupyter-js-services';
+
+import {
+  IDisposable, DisposableDelegate
+} from 'phosphor-disposable';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+
+/**
+ * The interface for a document model.
+ */
+export interface IDocumentModel {
+  /**
+   * Serialize the model.
+   */
+  serialize(): string;
+
+  /**
+   * Deserialize the model from a string.
+   */
+  deserialize(value: string): void;
+
+  /**
+   * The default kernel name for the the document.
+   */
+  defaultKernelName: string;
+
+  /**
+   * The default kernel language for the document.
+   */
+  defaultKernelLanguage: string;
+}
+
+
+/**
+ * A session info object for running sessions.
+ */
+export interface ISessionInfo {
+  /**
+   * The list of file paths associated with the running sessions.
+   */
+  paths: string[];
+
+  /**
+   * The list of kernel specs associated with the running sessions.
+   */
+  specs: IKernelSpecId[];
+}
+
+
+/**
+ * The document context object.
+ */
+export interface IDocumentContext {
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  kernelChanged: ISignal<IDocumentContext, IKernel>;
+
+  /**
+   * The current kernel associated with the document.
+   */
+  kernel: IKernel;
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: IKernelId): Promise<IKernel>;
+
+  /**
+   * Save the the document contents to disk.
+   */
+  save(): Promise<void>;
+
+  /**
+   * Revert the document contents to disk contents.
+   */
+  revert(): Promise<void>;
+
+  /**
+   * Get the list of running sessions.
+   */
+  listRunningSessions(): ISessionInfo;
+
+  /**
+   * Add a sibling widget to the document manager.
+   *
+   * @param widget - The widget to add to the document manager.
+   *
+   * @returns A disposable used to remove the sibling if desired.
+   */
+  addSibling(widget: Widget): IDisposable;
+}
+
+
+/**
+ * The interface for a widget factory.
+ */
+export
+interface IWidgetFactory<T extends Widget> {
+  /**
+   * Create a new widget given a document model and a context.
+   */
+  createNew(model: IDocumentModel, context: IDocumentContext): T;
+
+  /**
+   * Get the default kernel name given a list of specs.
+   */
+  getDefaultKernel(specs: IKernelSpecId[]): string;
+
+  /**
+   * Get the preferred kernel names given a list of specs.
+   */
+  getPreferredKernel(specs: IKernelSpecId[]): string[];
+
+  /**
+   * Get the preferred widget title given a path.
+   */
+  getWidgetTitle(path: string): string;
+}
+
+
+/**
+ * The options uses to register a widget factory.
+ */
+export
+interface IWidgetFactoryOptions {
+  /**
+   * The type of file the widget can view.
+   */
+  fileType: string;
```

Comment:
Yes, I think so. We definitely want to support registering the same widgetfactory for multiple filetypes.